src/app-layer-ftp.c

```diff
@@ -825,40 +730,49 @@ static AppLayerResult FTPParseResponse(Flow *f, void *ftp_state, AppLayerParserS
                 state->active = true;
                 tx->dyn_port = state->dyn_port;
                 tx->active = true;
-                SCLogDebug("FTP active mode (v4): dynamic port %"PRIu16"", dyn_port);
+                SCLogDebug("FTP active mode (v4): dynamic port %" PRIu16 "", dyn_port);
                 break;
 
             case FTP_COMMAND_PASV:
-                if (state->current_line_len >= 4 && SCMemcmp("227 ", state->current_line, 4) == 0) {
-                    FTPParsePassiveResponse(f, ftp_state, state->current_line, state->current_line_len);
+                if (line.len >= 4 && SCMemcmp("227 ", line.buf, 4) == 0) {
+                    FTPParsePassiveResponse(f, ftp_state, line.buf, line.len);
                 }
                 break;
 
             case FTP_COMMAND_EPSV:
-                if (state->current_line_len >= 4 && SCMemcmp("229 ", state->current_line, 4) == 0) {
-                    FTPParsePassiveResponseV6(f, ftp_state, state->current_line, state->current_line_len);
+                if (line.len >= 4 && SCMemcmp("229 ", line.buf, 4) == 0) {
+                    FTPParsePassiveResponseV6(f, ftp_state, line.buf, line.len);
                 }
                 break;
             default:
                 break;
         }
 
-        if (likely(state->current_line_len)) {
+        if (likely(line.len)) {
             FTPString *response = FTPStringAlloc();
             if (likely(response)) {
-                response->len = CopyCommandLine(&response->str, state->current_line, state->current_line_len);
+                response->len = CopyCommandLine(&response->str, &line);
                 response->truncated = state->current_line_truncated;
                 TAILQ_INSERT_TAIL(&tx->response_list, response, next);
             }
         }
 
         /* Handle preliminary replies -- keep tx open */
-        if (FTPIsPPR(state->current_line, state->current_line_len)) {
-            continue;
+        if (FTPIsPPR(line.buf, line.len)) {
+            goto call_getline;
         }
     tx_complete:
         tx->done = true;
+
+        if (line.len >= ftp_max_line_len) {
+            ftpi.consumed = ftpi.len + 1;
+            break;
+        }
+    call_getline:
```

Comment:
can we do this loop w/o goto's? Might be cleaner to factor out some parts of the loop into util funcs